packages/versioning/src/lib.rs

```diff
@@ -1,115 +1,180 @@
+            "Couldn't upgrade contract because storage version didn't match expected one!",
+        )
+        .into());
     }
 
-    const VERSION: WithVersion<VariantsQuery> = WithVersion::Version { version: () };
+    CW_VERSION_ITEM.remove(storage);
 
-    const QUERY_ABC: WithVersion<VariantsQuery> = WithVersion::new_query(VariantsQuery::Abc {});
+    OLD_VERSION_ITEM.remove(storage);
 
-    const QUERY_DEF: WithVersion<VariantsQuery> = WithVersion::new_query(VariantsQuery::Def {});
+    // Using zero as a starting storage version to mark this as a new epoch.
+    initialize(storage, version)?;
 
-    const QUERY_VERSION: WithVersion<VariantsQuery> =
-        WithVersion::new_query(VariantsQuery::Version {});
+    migrate_storage_functor.map_or(Ok(()), move |functor| functor(storage))
+}
 
-    fn assert_query_serde(value: WithVersion<VariantsQuery>) {
-        assert_eq!(
-            serde_json::from_str::<WithVersion<VariantsQuery>>(
-                &serde_json::to_string(&value).unwrap()
-            )
-            .unwrap(),
-            value
-        );
+#[inline]
+pub fn update_software(storage: &mut dyn Storage, version: Version) -> StdResult<()> {
+    update_version(storage, version.storage, version).map(|_| ())
+}
+
+pub fn update_software_and_storage<
+    'r,
+    const FROM_STORAGE_VERSION: VersionSegment,
+    MigrateStorageFunctor,
+    MigrateStorageError,
+>(
+    storage: &'r mut dyn Storage,
+    version: Version,
+    migrate_storage: MigrateStorageFunctor,
+) -> Result<(), MigrateStorageError>
+where
+    MigrateStorageFunctor: FnOnce(&'r mut dyn Storage) -> Result<(), MigrateStorageError>,
+    MigrateStorageError: From<StdError> + Error,
+{
+    if version.storage == FROM_STORAGE_VERSION {
+        return Err(StdError::generic_err("Software and storage update handler called, but expected and new storage versions are the same!").into());
     }
 
-    #[test]
-    fn test_query_serde() {
-        assert_query_serde(VERSION);
-        assert_query_serde(QUERY_ABC);
-        assert_query_serde(QUERY_DEF);
-        assert_query_serde(QUERY_VERSION);
-
-        assert_eq!(
-            serde_json::from_str::<WithVersion<VariantsQuery>>(r#"{"version":null}"#).unwrap(),
-            VERSION
-        );
-
-        assert_eq!(
-            serde_json::from_str::<WithVersion<VariantsQuery>>(r#"{"abc":{}}"#).unwrap(),
-            QUERY_ABC
-        );
-
-        assert_eq!(
-            serde_json::from_str::<WithVersion<VariantsQuery>>(r#"{"def":{}}"#).unwrap(),
-            QUERY_DEF
-        );
-
-        assert_eq!(
-            serde_json::from_str::<WithVersion<VariantsQuery>>(r#"{"version":{}}"#).unwrap(),
-            QUERY_VERSION
-        );
+    if version.storage != FROM_STORAGE_VERSION.wrapping_add(1) {
+        return Err(StdError::generic_err("Expected and new storage versions are not directly adjacent! This could indicate an error!").into());
     }
+
+    update_version(storage, version.storage, version)?;
```

Comment:
should not pass `FROM_STORAGE_VERSION` instead?

Reply:
Yes, it should be.